File: src/main/java/gargoyle/ct/AbstractResource.java
package gargoyle.ct;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URISyntaxException;
import java.net.URL;
import java.net.URLConnection;

abstract class AbstractResource implements Resource {
    private static final String USER_AGENT = "Mozilla/5.0 (Windows; U; Windows NT 6.0; en-US; rv:1.9.1.2) Gecko/20090729 Firefox/3.5.2 (.NET CLR 3.5.30729)";
    private static final String PROP_UA = "User-Agent";
    private static final String METHOD_HEAD = "HEAD";
    private final String location;

    protected AbstractResource(final String location) {
        this.location = location;
    }

    protected abstract <R extends Resource> R createResource(final R base, final String location);

    @Override
    public boolean exists() {
        try {
            return this.exists(this.toURL());
        } catch (final IOException ex) {
            return false;
        }
    }

    protected boolean exists(final URL url) {
        if (url != null) {
            try {
                final URLConnection connection = url.openConnection();
                if (connection instanceof HttpURLConnection) {
                    final HttpURLConnection huc = (HttpURLConnection) connection;
                    huc.setInstanceFollowRedirects(false);
                    huc.setRequestMethod(AbstractResource.METHOD_HEAD);
                    huc.setRequestProperty(AbstractResource.PROP_UA, AbstractResource.USER_AGENT);
                    huc.connect();
                    return huc.getResponseCode() == HttpURLConnection.HTTP_OK;
                }
                if ("file".equalsIgnoreCase(url.getProtocol())) {
                    return new File(url.toURI()).exists();
                }
            } catch (final IOException ex) {
                return false;
            } catch (final URISyntaxException ex) {
                return false;
            }
            return false;
        }
        return false;
    }

    @Override
    public InputStream getInputStream() throws IOException {
        return this.toURL().openConnection().getInputStream();
    }

    @Override
    public String getLocation() {
        return this.location;
    }

    @Override
    public OutputStream getOutputStream() throws IOException {
        return this.toURL().openConnection().getOutputStream();
    }

    @Override
    public String toString() {
        return "AbstractResource [location=" + this.location + "]";
    }

    @Override
    public URL toURL() throws IOException {
        return new URL(this.location);
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/resource/internal/AbstractResource.kt
package gargoyle.ct.util.resource.internal

import gargoyle.ct.util.resource.Resource
import gargoyle.ct.util.util.simpleClassName
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.URL

/**
 * base functionality for [Resource]
 */
abstract class AbstractResource<R : AbstractResource<R>> protected constructor(override val location: String) :
    AbstractResourceBase(),
    Resource {

    protected abstract fun createResource(base: R?, location: String): R
    override fun exists(): Boolean =
        try {
            exists(toURL())
        } catch (ex: Exception) {
            false
        }

    override val inputStream: InputStream
        get() = getInputStream(toURL())

    override val outputStream: OutputStream
        get() = getOutputStream(toURL(), true)
    override val isBase: Boolean
        get() = true
    override val isReadable: Boolean
        get() = try {
            isReadable(toURL())
        } catch (e: IOException) {
            false
        }
    override val isWritable: Boolean
        get() = try {
            isWritable(toURL())
        } catch (e: IOException) {
            false
        }

    override fun toURL(): URL = URL(location)

    override fun toString(): String = "${this::class.simpleClassName} [location=${location}]"
}
